cpython/search_pypi_top.py

```diff
@@ -138,39 +134,76 @@ def is_binary_file(args, fp):
         return True
     return False
 
+
 def grep(args, archive_filename, regex):
     for filename, fp in decompress(args, archive_filename):
         if is_binary_file(args, fp):
-            logging.info(f"ignore binary file: {archive_filename}: {filename}")
+            logger.info(f"ignore binary file: {archive_filename}: {filename}")
             continue
 
-        matchs = []
+        matches = []
         ignore = False
         lineno = 1
         # Split at Unix newline b'\n' byte
         for line in fp:
             if lineno == 1 and IGNORE_CYTHON and CYTHON_REGEX.match(line):
-                logging.info(f"ignore Cython file: {archive_filename}: {filename}")
+                logger.info(f"ignore Cython file: {archive_filename}: {filename}")
                 ignore = True
                 break
             if regex.search(line):
-                matchs.append((filename, line))
+                matches.append((filename, line))
             lineno += 1
 
-        if matchs and not ignore:
-            yield from matchs
+        if matches and not ignore:
+            yield from matches
+
+
+def search_file(filename, index, len_filenames, args, pypi_dir, regex):
+    lines = 0
+    results = []
+    filename = os.path.join(pypi_dir, filename)
+    percent = index * 100 / len_filenames
+    logger.warning(f"grep {filename} ({percent:.0f}%, {index}/{len_filenames})")
+
+    for name, line in grep(args, filename, regex):
+        line = line.decode('utf8', 'replace').strip()
+
+        result = f"{filename}: {name}: {line}"
+        print(result, flush=True)
+        results.append(result)
+        lines += 1
+
+    return lines, filename if lines >= 1 else None, results
 
 
 def search_dir(args, pypi_dir, pattern):
+    lines = 0
+    projects = set()
+    all_results = []
     regex = re.compile(pattern)
     filenames = os.listdir(pypi_dir)
-    for index, filename in enumerate(filenames, 1):
-        filename = os.path.join(pypi_dir, filename)
-        percent = index * 100 / len(filenames)
-        logging.warning(f"grep {filename} ({percent:.0f}%, {index}/{len(filenames)})")
-        for name, line in grep(args, filename, regex):
-            line = line.decode('utf8', 'replace').strip()
-            yield (filename, name, line)
+
+    with multiprocessing.Pool() as pool:
+        ret = pool.starmap(
+            search_file,
+            zip(
+                filenames,
+                range(len(filenames)),
+                repeat(len(filenames)),
+                repeat(args),
+                repeat(pypi_dir),
+                repeat(regex),
+            )
+        )
+        for new_lines, new_project, results in ret:
+            lines += new_lines
+            if new_project:
+                projects.add(new_project)
+            all_results.extend(results)
+    pool.close()
+    pool.join()
```

Comment:
When I interrupt the job with CTRL+C, I get messages like:

```
^CProcess ForkPoolWorker-1:
Process ForkPoolWorker-3:
Process ForkPoolWorker-8:
Process ForkPoolWorker-2:
Process ForkPoolWorker-5:
Process ForkPoolWorker-4:
Process ForkPoolWorker-6:
Process ForkPoolWorker-7:
```

I don't know where they come from.